<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>R package containing a suite of functions for single-cell RNA-Seq data post-processing, involving unsupervised clustering, dimensionality reduction, batch correction, sample-cluster composition analysis, pseudobulk differential expression, and functional annotation ‚Ä¢ scpost</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="R package containing a suite of functions for single-cell RNA-Seq data post-processing, involving unsupervised clustering, dimensionality reduction, batch correction, sample-cluster composition analysis, pseudobulk differential expression, and functional annotation">
<meta property="og:description" content="This R package enables the post-processing of single-cell RNA-Seq data, involving unsupervised clustering, dimensionality reduction, batch correction, sample-cluster composition analysis, pseudobulk differential expression, and functional annotation.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scpost</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette_scpost_usage.html">scpost: Post-Processing Guide for Single-cell Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="float">
<img src="inst/extdata/scpost_Logo.png" alt="Logo"><div class="figcaption">Logo</div>
</div>
<hr>
<div class="section level1">
<div class="page-header"><h1 id="scpost">
<strong>scpost</strong><a class="anchor" aria-label="anchor" href="#scpost"></a>
</h1></div>
<p>An R package for post-processing single-cell RNA-Seq data involving unsupervised clustering, dimensionality reduction, batch correction, sample-cluster composition calculation, pseudobulk differential expression (in progress), and functional annotation (in progress).</p>
<p>This package is designed to process data aggregated by <strong><a href="https://github.com/g-duclos/scprep" class="external-link">scprep</a></strong> and now supports all three output object types: <strong>ExpressionSet</strong>, <strong>Seurat</strong>, and <strong>SingleCellExperiment</strong>.</p>
<p>Data that may be used for testing <strong>scpost</strong> may be accessed via the <strong><a href="https://github.com/g-duclos/scdata" class="external-link">scdata</a></strong> package.</p>
<hr>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="option-1-install-from-github">Option 1: Install from GitHub<a class="anchor" aria-label="anchor" href="#option-1-install-from-github"></a>
</h3>
<p>The <strong>scpost</strong> R package can be installed from Github using devtools:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"g-duclos/scpost"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-2-docker-container">Option 2: Docker Container<a class="anchor" aria-label="anchor" href="#option-2-docker-container"></a>
</h3>
<p>For a containerized environment with R, <strong>scpost</strong>, and all dependencies pre-installed:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Clone the repository</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/g-duclos/scpost.git</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># Navigate to directory with cloned 'scpost'</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="bu">cd</span> scpost</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># Build the Docker image</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="ex">docker</span> build <span class="at">-t</span> scpost .</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># Run interactive R session with scpost</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">--rm</span> scpost</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>üìñ <strong><a href="https://g-duclos.github.io/scpost/articles/vignette_scpost_usage.html" class="external-link">Complete Usage Guide</a></strong> - Guide demonstrating post-processing workflows for all object types</p>
<hr>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level4">
<h4 id="parameters">Parameters<a class="anchor" aria-label="anchor" href="#parameters"></a>
</h4>
<p>Specify the pipeline parameters - view the parameters file here: <a href="inst/extdata/scpost_parameters.csv">scpost_parameters.csv</a> (click ‚ÄúRaw file‚Äù at the top right to download)</p>
<p><strong>Critical Parameters</strong> * <em>dir_input</em> (path to input directory containing output from scprep) * <em>dir_output</em> (path to output directory) * <em>batch_correction_method</em> (FALSE, ‚Äúharmony‚Äù, or ‚Äúcca‚Äù) * <em>batch_var</em> (metadata column name for batch variable, or FALSE) * <em>pc_max</em> (maximum number of principal components to use) * <em>res_ops</em> (comma-separated clustering resolution values)</p>
</div>
<div class="section level4">
<h4 id="input-directory">Input Directory<a class="anchor" aria-label="anchor" href="#input-directory"></a>
</h4>
<p>Define the input directory (<em>dir_input</em>), which must contain one of the following files generated by the <strong><a href="https://github.com/g-duclos/scprep" class="external-link">scprep</a></strong> R package:</p>
<p><strong>Supported Input Files:</strong> * <strong>ExpressionSet.rds</strong> (Bioconductor ExpressionSet object with GEX counts + metadata) * <strong>seurat.rds</strong> (Seurat v5 object with GEX counts + metadata) * <strong>sce.rds</strong> (SingleCellExperiment object with GEX counts + metadata)</p>
<p><strong>Note:</strong> The object type is automatically detected based on which file is present in the input directory.</p>
</div>
<div class="section level4">
<h4 id="dependencies">Dependencies:<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h4>
<p><strong>scpost</strong> requires the following R packages: - <strong>Seurat</strong> (&gt;= 5.0.0): For single-cell analysis workflows and data processing - <strong>harmony</strong> (optional): For Harmony batch correction method - <strong>Biobase</strong>: For ExpressionSet data structures - <strong>SingleCellExperiment</strong>: For SingleCellExperiment object support - <strong>SummarizedExperiment</strong>: For SCE data manipulation - <strong>S4Vectors</strong>: For metadata operations - <strong>R</strong> (&gt;= 4.0.0): Minimum R version</p>
<p><strong>Note:</strong> Seurat is NOT included as a formal package dependency due to common installation complications, but must be installed separately. All other dependencies are automatically installed when using <code><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">devtools::install_github()</a></code>.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="supported-object-types">Supported Object Types<a class="anchor" aria-label="anchor" href="#supported-object-types"></a>
</h2>
<p><strong>scpost</strong> now supports three different input/output object types, maintaining compatibility with data generated by <strong>scprep</strong>:</p>
<div class="section level3">
<h3 id="expressionset-original">
<strong>ExpressionSet</strong> (Original)<a class="anchor" aria-label="anchor" href="#expressionset-original"></a>
</h3>
<ul>
<li>Classic Bioconductor S4 object for storing expression data</li>
<li>Results stored in <code>assayData$Seurat</code> and <code><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData()</a></code> slots</li>
<li>Compatible with traditional Bioconductor workflows</li>
</ul>
</div>
<div class="section level3">
<h3 id="seurat-v5">
<strong>Seurat</strong> (v5)<a class="anchor" aria-label="anchor" href="#seurat-v5"></a>
</h3>
<ul>
<li>Popular single-cell analysis object</li>
<li>Results stored in <code>@reductions</code>, <code>@misc</code>, and <code>@meta.data</code> slots</li>
<li>Seamlessly integrates with existing Seurat-based pipelines</li>
</ul>
</div>
<div class="section level3">
<h3 id="singlecellexperiment">
<strong>SingleCellExperiment</strong><a class="anchor" aria-label="anchor" href="#singlecellexperiment"></a>
</h3>
<ul>
<li>Modern Bioconductor S4 object for single-cell data</li>
<li>Results stored in <code>reducedDims()</code>, <code>metadata()</code>, and <code>colData()</code> slots</li>
<li>Compatible with Bioconductor/scater workflows</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Template function to perform unsupervised clustering, dimensionality reduction, batch correction, sample-cluster composition calculation, pseudobulk differential expression (in progress), and functional annotation (in progress).</p>
<p><strong>Works with all three object types:</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Automatically detects object type and processes accordingly</span></span>
<span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="fu">scpost</span><span class="fu">::</span><span class="fu"><a href="reference/template_scpost.html">template_scpost</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="va">dir_output</span><span class="op">)</span></span></code></pre></div>
<p><strong>Core functions:</strong></p>
<details><summary>
Build or process Seurat object with normalization, highly variable gene selection, scaling, and principal component analysis (PCA).
</summary><pre>
# For ExpressionSet or SingleCellExperiment inputs:
# Select high quality cells
cells 
<pre>
# Select genes with minimum expression requirements (ExpressionSet only)
genes 
<pre>
# Build new Seurat object from counts matrix (for ExpressionSet/SCE)
seurat.obj 
<pre>
# For Seurat inputs:
# Existing Seurat object is processed directly (normalize ‚Üí scale ‚Üí PCA)
# No need to create new object - original object is enriched with results
</pre>
</pre></pre></pre></details><details><summary>
Data integration intended to correct for technical batch effects for downstream clustering &amp; dimensionality reduction, based on a selected ‚Äúbatch‚Äù variable of interest using one of the following methods:
<ul>
<li>
<a href="https://github.com/immunogenomics/harmony" class="external-link">Harmony R package</a>
</li>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>    <span class="sc">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>        <span class="er">&lt;</span>a href<span class="ot">=</span><span class="st">"https://satijalab.org/seurat/articles/integration_rpca.html"</span><span class="sc">&gt;</span>Seurat<span class="st">'s CCA/RPCA method&lt;/a&gt;</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="st">    &lt;/li&gt;</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="st">&lt;/summary&gt;</span></span></code></pre></div>
</ul>
<pre>
# Extract batch metadata from appropriate object type
# For ExpressionSet:
batch_metadata 
<pre>
# Perform batch correction with "harmony" or "cca" method
seurat.obj 
</pre></pre></summary></details><details><summary>
A single command to perform unsupervised cell clustering using Seurat‚Äôs standard methodology (SNN &amp; Louvain community detection) with the option of returning a range of cluster solutions.
</summary><pre>
# Perform unsupervised clustering at specified resolution
seurat.obj 
</pre></details><details><summary>
The Seurat <em>RunTSNE</em> function is used to run t-SNE on the dataset.
</summary><ul>
<li>
tSNE results are stored in object-specific locations:
<ul>
<li>
<strong>ExpressionSet:</strong> assayData<span class="math inline">$Seurat[["tSNE"]]&lt;/li&gt;
            &lt;li&gt;&lt;strong&gt;Seurat:&lt;/strong&gt; @misc$</span>tSNE_iterations
</li>
<li>
<strong>SingleCellExperiment:</strong> reducedDim(, ‚ÄútSNE_Iteration_X‚Äù)
</li>
</ul>
</li>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="sc">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>    tSNE results will reflect the decision to use or not use one of the specified batch correction methods</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="sc">&lt;</span><span class="er">/</span>li<span class="sc">&gt;</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="er">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>    According to the number of iterations specified <span class="cf">in</span> the <span class="st">"scpost_parameters.csv"</span> file, results are subsequently stored as <span class="st">"Iteration_1"</span>, <span class="st">"Iteration_2"</span>, ...</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="sc">&lt;</span><span class="er">/</span>li<span class="sc">&gt;</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="er">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    According to the number of iterations specified <span class="cf">in</span> the <span class="st">"scpost_parameters.csv"</span> file, results are saved <span class="cf">in</span> a subdirectory named <span class="st">"scpost_analysis"</span> <span class="cf">in</span> <span class="sc">*</span>dir_output<span class="sc">*</span> <span class="cf">in</span> the RDS format as <span class="st">"tSNE_Iteration_1.rds"</span>, <span class="st">"tSNE_Iteration_2.rds"</span>, ...</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="sc">&lt;</span><span class="er">/</span>li<span class="sc">&gt;</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="er">&lt;/</span>ul<span class="sc">&gt;</span></span></code></pre></div>
</ul></details><details><summary>
The Seurat <em>RunUMAP</em> function is used to run UMAP on the dataset.
</summary><ul>
<li>
UMAP results are stored in object-specific locations:
<ul>
<li>
<strong>ExpressionSet:</strong> assayData<span class="math inline">$Seurat[["UMAP"]]&lt;/li&gt;
            &lt;li&gt;&lt;strong&gt;Seurat:&lt;/strong&gt; @misc$</span>UMAP_iterations
</li>
<li>
<strong>SingleCellExperiment:</strong> reducedDim(, ‚ÄúUMAP_Iteration_X‚Äù)
</li>
</ul>
</li>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>    UMAP results will reflect the decision to use or not use one of the specified batch correction methods</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="sc">&lt;</span><span class="er">/</span>li<span class="sc">&gt;</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="er">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    According to the number of iterations specified <span class="cf">in</span> the <span class="st">"scpost_parameters.csv"</span> file, results are subsequently stored as <span class="st">"Iteration_1"</span>, <span class="st">"Iteration_2"</span>, ...</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="sc">&lt;</span><span class="er">/</span>li<span class="sc">&gt;</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="er">&lt;</span>li<span class="sc">&gt;</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    According to the number of iterations specified <span class="cf">in</span> the <span class="st">"scpost_parameters.csv"</span> file, results are saved <span class="cf">in</span> a subdirectory named <span class="st">"scpost_analysis"</span> <span class="cf">in</span> <span class="sc">*</span>dir_output<span class="sc">*</span> <span class="cf">in</span> the RDS format as <span class="st">"UMAP_Iteration_1.rds"</span>, <span class="st">"UMAP_Iteration_2.rds"</span>, ...</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="sc">&lt;</span><span class="er">/</span>li<span class="sc">&gt;</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="er">&lt;/</span>ul<span class="sc">&gt;</span></span></code></pre></div>
</ul></details><li>
Sample-cluster composition assessment methods with built-in statistical methods for univariate or multivariate analysis of cluster-associations with variable(s) of interest are a work in progress ‚Ä¶
</li>
<p><br></p>
<li>
Pseudobulk differential expression functionalities, involving the use of the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html" class="external-link">DESeq2 R package</a> are a work in progress‚Ä¶
</li>
<p><br></p>
<li>
Functional annotation, involving the use of gene set databases, including <a href="https://www.gsea-msigdb.org/gsea/msigdb/" class="external-link">the Molecular Signatures Database (MSigDB)</a> and <a href="https://geneontology.org" class="external-link">Gene Ontology (GO)</a>, with methods, including the <a href="https://bioconductor.org/packages/release/bioc/html/fgsea.html" class="external-link">Fast Gene Set Enrichment analysis (fgsea) R package</a> and the <a href="https://cran.r-project.org/web/packages/enrichR/index.html" class="external-link">enrichR R package</a>
</li>
<p><br></p>
<details><summary>
Additional features
</summary><ul>
<li>
Automatically detects input object type (ExpressionSet, Seurat, or SingleCellExperiment) and processes accordingly
</li>
<li>
Parameters specified in scpost_parameters.csv are stored in object-specific locations:
<ul>
<li>
<strong>ExpressionSet:</strong> assayData<span class="math inline">$Params[["scpost_Parameters"]]&lt;/li&gt;
        &lt;li&gt;&lt;strong&gt;Seurat:&lt;/strong&gt; @misc$</span>scpost_Parameters
</li>
<li>
<strong>SingleCellExperiment:</strong> metadata()$scpost_Parameters
</li>
</ul>
</li>
<li>
Random seeds for reproducible UMAP/tSNE iterations are either retrieved from the input object (if generated by scprep) or automatically generated and stored
</li>
<li>
Clustering results stored in appropriate metadata slots:
<ul>
<li>
<strong>ExpressionSet:</strong> pData()<span class="math inline"><em>S</em><em>e</em><em>u</em><em>r</em><em>a</em><em>t</em><sub><em>C</em></sub><em>l</em><em>u</em><em>s</em><em>t</em><em>e</em><em>r</em><em>s</em><sub><em>R</em></sub><em>e</em><em>s</em>‚ÄÖ‚àí‚ÄÖ<em>X</em>‚ÄÑ&lt;‚ÄÑ/<em>l</em><em>i</em>‚ÄÑ&gt;‚ÄÑ‚ÄÑ&lt;‚ÄÑ<em>l</em><em>i</em>‚ÄÑ&gt;‚ÄÑ‚ÄÑ&lt;‚ÄÑ<em>s</em><em>t</em><em>r</em><em>o</em><em>n</em><em>g</em>‚ÄÑ&gt;‚ÄÑ<em>S</em><em>e</em><em>u</em><em>r</em><em>a</em><em>t</em>‚ÄÑ:‚ÄÑ‚ÄÑ&lt;‚ÄÑ/<em>s</em><em>t</em><em>r</em><em>o</em><em>n</em><em>g</em>‚ÄÑ&gt;‚ÄÑ@<em>m</em><em>e</em><em>t</em><em>a</em>.<em>d</em><em>a</em><em>t</em><em>a</em></span>Seurat_Clusters_Res-X
</li>
<li>
<strong>SingleCellExperiment:</strong> colData()$Seurat_Clusters_Res-X
</li>
</ul>
</li>
<li>
PCA and batch-corrected embeddings stored in object-specific reduction slots:
<ul>
<li>
<strong>ExpressionSet:</strong> assayData<span class="math inline">$Seurat[["PCA"]] and assayData$</span>Seurat[[‚ÄúPCA_Harmony‚Äù]]
</li>
<li>
<strong>Seurat:</strong> @reductions<span class="math inline"><em>p</em><em>c</em><em>a</em><em>a</em><em>n</em><em>d</em>@<em>r</em><em>e</em><em>d</em><em>u</em><em>c</em><em>t</em><em>i</em><em>o</em><em>n</em><em>s</em></span>harmony
</li>
<li>
<strong>SingleCellExperiment:</strong> reducedDim(, ‚ÄúPCA‚Äù) and reducedDim(, ‚ÄúPCA_Harmony‚Äù)
</li>
</ul>
</li>
<li>
Output object saved with appropriate filename:
<ul>
<li>
<strong>ExpressionSet:</strong> ExpressionSet.rds
</li>
<li>
<strong>Seurat:</strong> seurat.rds
</li>
<li>
<strong>SingleCellExperiment:</strong> sce.rds
</li>
</ul>
</li>
<li>
All intermediate results (PCA embeddings, clustering metadata, UMAP/tSNE iterations) saved as separate .rds files in scpost_analysis subdirectory for easy downstream access
</li>
</ul></details><hr>
</div>
<div class="section level2">
<h2 id="data-flow-scprep--scpost">Data Flow: scprep ‚Üí scpost<a class="anchor" aria-label="anchor" href="#data-flow-scprep--scpost"></a>
</h2>
<p><strong>scpost</strong> seamlessly integrates with <strong>scprep</strong> output, supporting all three object types:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Workflow 1: ExpressionSet (Original)</span></span>
<span><span class="fu">scprep</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scprep/man/template_scprep.html" class="external-link">template_scprep</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="st">"/path"</span><span class="op">)</span>  <span class="co"># output_type="eset"</span></span>
<span><span class="fu">scpost</span><span class="fu">::</span><span class="fu"><a href="reference/template_scpost.html">template_scpost</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="st">"/path"</span><span class="op">)</span>  <span class="co"># Reads ExpressionSet.rds</span></span>
<span></span>
<span><span class="co"># Workflow 2: Seurat (New)</span></span>
<span><span class="fu">scprep</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scprep/man/template_scprep.html" class="external-link">template_scprep</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="st">"/path"</span><span class="op">)</span>  <span class="co"># output_type="seurat"</span></span>
<span><span class="fu">scpost</span><span class="fu">::</span><span class="fu"><a href="reference/template_scpost.html">template_scpost</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="st">"/path"</span><span class="op">)</span>  <span class="co"># Reads seurat.rds</span></span>
<span></span>
<span><span class="co"># Workflow 3: SingleCellExperiment (New)</span></span>
<span><span class="fu">scprep</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scprep/man/template_scprep.html" class="external-link">template_scprep</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="st">"/path"</span><span class="op">)</span>  <span class="co"># output_type="sce"</span></span>
<span><span class="fu">scpost</span><span class="fu">::</span><span class="fu"><a href="reference/template_scpost.html">template_scpost</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="st">"/path"</span><span class="op">)</span>  <span class="co"># Reads sce.rds</span></span></code></pre></div>
<p><strong>Note:</strong> The <code>output_type</code> parameter is specified in <code>scprep_parameters.csv</code> when running <strong>scprep</strong>. <strong>scpost</strong> automatically detects and processes the corresponding object type.</p>
<hr>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>use_gpl3_license()</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scpost</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Grant Duclos <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/YOUR-ORCID-ID" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Grant Duclos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
